### DISCUSSION TITLE
Http Clients Pools Handling

### DISCUSSION THREAD

**Discussion by shalevamit9** [nodejs/undici#2773] [Q&A] [answered]
I see in npm docs and GitHub as well the common methods like:

- request
- stream
- pipeline
- connect
- fetch

The question is, for example the simple example of:
```js
const { request } = require('undici')
async function getRequest (port = 3001) {
  // A simple GET request
  const {
    statusCode,
    headers,
    body
  } = await request(`http://localhost:${port}/`)

  const data = await body.text()
  console.log('response received', statusCode)
  console.log('headers', headers)
  console.log('data', data)
}
```

1. Does the `undici` package handle the client creation, reuse, and manage pool of clients when using the simple `request` method?
I'm asking this as I don't see any hint regarding this in docs. I only see that there's client and clients pool offered by `undici` but is it managed behind the scenes?
2. Another question if it does handle a connection pool behind the scenes is, how does it handle DNS invalidation?

**Comment by metcoder95** [comment]
👋 
1. By default, it creates a Pool of connections to handle multiple `origin`
2. It does not; once the DNS resolution is done, the connection will prevail until it goes `idle` and later dropped or dropped manually for either side of the connection. After that, a new DNS resolution will be done, and from there things like host DNS caching and others plays its part.
	- We are seeking to implement DNS caching, but that still WIP

**Reply by shalevamit9** [reply]
Hey @metcoder95, thanks for the answer, the first point is very clear. I want to clarify a scenario if it's possible regarding the second point.
Let's take for example the following simple express app that uses `undici`:
```js
const express = require("express");
const { request } = require("undici");

const app = express();

app.get("/", async (req, res) => {
  const response = await request("http://some-endpoints");

  await response.body.dump();

  res.end();
});

app.listen(3000, () => console.log("up"));
```
Let's say that this single endpoint keeps getting called without stop and some time after creating the connection, `some-endpoints` DNS changes. At the moment I have two ideas:

1. Recognize that sort of error and dispose the connection manually.
2. Minimize the `idle` time so it will dispose automatically.

Am I at the right direction for a current solution?

**Reply by metcoder95** [reply, marked as answer]
Yes, that approach is correct and the only way to sort that out for the moment. If the DNS changes and the IP comes with it, `undici` will just fail because the connection might be dropped or not available anymore. Then a new reconnect will be necessary.

As you suggest, you can lower the `keepAliveMaxTimeout` to reduce the life of the connection and drop it faster (https://undici.nodejs.org/#/docs/api/Client?id=parameter-clientoptions), if you want to have a faster cycle to minimize this possibility.
